src/rosdistro/manifest_provider/git.py

```diff
@@ -47,102 +49,76 @@
 def git_manifest_provider(_dist_name, repo, pkg_name):
     assert repo.version
     try:
-        release_tag = get_release_tag(repo, pkg_name)
-        package_xml = _get_package_xml(repo.url, release_tag)
-        return package_xml
+        release_tag = repo.get_release_tag(pkg_name)
+        with _temp_git_clone(repo.url, release_tag) as git_repo_path:
+            filename = os.path.join(git_repo_path, 'package.xml')
+            if not os.path.exists(filename):
+                raise RuntimeError('Could not find package.xml in repository "%s"' % repo.url)
+            with open(filename, 'r') as f:
+                return f.read()
     except Exception as e:
         raise RuntimeError('Unable to fetch package.xml: %s' % e)
 
 
-def _git_version_gte(version):
-    global _git_client_version
-    if not _git_client_version:
-        cmd = [_git_client_executable, '--version']
-        result = _run_command(cmd)
-        _git_client_version = result['output'].split()[-1]
-    return LooseVersion(_git_client_version) >= LooseVersion(version)
+def git_source_manifest_provider(repo):
+    try:
+        with _temp_git_clone(repo.url, repo.version) as git_repo_path:
+            # Include the git hash in our cache dictionary.
+            result = Git(git_repo_path).command('rev-parse', 'HEAD')
+            cache = { '_ref': result['output'] }
+
+            for path, dirnames, filenames in os.walk(git_repo_path):
+                # Don't recurse into hidden folders.
+                if os.path.basename(path)[0] == '.':
+                    dirnames[:] = []
+                    continue
+
+                if 'package.xml' in filenames:
+                    filename = os.path.join(path, 'package.xml')
+                    with open(filename, 'r') as f:
+                        package_xml = f.read()
+
+                    try:
+                        name = parse_package_string(package_xml).name
+                    except InvalidPackage:
+                        raise RuntimeError('Unable to parse package.xml file found in %s' % repo.url)
+                    cache[name] = [ path[len(git_repo_path)+1:], package_xml ]
+
+                    # Prevent further recursion down this tree.
+                    dirnames[:] = []
+
+    except Exception as e:
+        raise RuntimeError('Unable to fetch source package.xml files: %s' % e)
+
+    return cache
 
 
-def _get_package_xml(url, tag):
+@contextmanager
+def _temp_git_clone(url, ref):
     base = tempfile.mkdtemp('rosdistro')
+    git = Git(base)
```

Comment:
Maybe pass `base` explicitly as a keyword argument. That makes the following code using `.` more clear:

```
git  = Git(cwd=base)
```